miniapps/solvers/lor_solvers.cpp

```diff
@@ -0,0 +1,200 @@
+
+   Mesh mesh(mesh_file, 1, 1);
+   int dim = mesh.Dimension();
+   MFEM_VERIFY(dim == 2 || dim == 3, "Spatial dimension must be 2 or 3.");
+   for (int l = 0; l < ref_levels; l++) { mesh.UniformRefinement(); }
+
+   FunctionCoefficient f_coeff(f), u_coeff(u);
+   VectorFunctionCoefficient f_vec_coeff(dim, f_vec), u_vec_coeff(dim, u_vec);
+
+   int b1 = BasisType::GaussLobatto, b2 = BasisType::Integrated;
+   unique_ptr<FiniteElementCollection> fec;
+   if (H1) { fec.reset(new H1_FECollection(order, dim, b1)); }
+   else if (ND) { fec.reset(new ND_FECollection(order, dim, b1, b2)); }
+   else if (RT) { fec.reset(new RT_FECollection(order-1, dim, b1, b2)); }
+   else { fec.reset(new L2_FECollection(order, dim, b1)); }
+
+   FiniteElementSpace fes(&mesh, fec.get());
+   cout << "Number of DOFs: " << fes.GetTrueVSize() << endl;
+
+   Array<int> ess_dofs;
+   // In DG, boundary conditions are enforced weakly, so no essential DOFs.
+   if (!L2) { fes.GetBoundaryTrueDofs(ess_dofs); }
+
+   BilinearForm a(&fes);
+   if (H1 || L2)
+   {
+      a.AddDomainIntegrator(new MassIntegrator);
+      a.AddDomainIntegrator(new DiffusionIntegrator);
+   }
+   else
+   {
+      a.AddDomainIntegrator(new VectorFEMassIntegrator);
+   }
+
+   if (ND) { a.AddDomainIntegrator(new CurlCurlIntegrator); }
+   else if (RT) { a.AddDomainIntegrator(new DivDivIntegrator); }
+   else if (L2)
+   {
+      a.AddInteriorFaceIntegrator(new DGDiffusionIntegrator(-1.0, kappa));
+      a.AddBdrFaceIntegrator(new DGDiffusionIntegrator(-1.0, kappa));
+   }
+   // TODO: L2 diffusion not implemented with partial assemble
+   if (!L2) { a.SetAssemblyLevel(AssemblyLevel::PARTIAL); }
+   a.Assemble();
+
+   LinearForm b(&fes);
+   if (H1 || L2) { b.AddDomainIntegrator(new DomainLFIntegrator(f_coeff)); }
+   else { b.AddDomainIntegrator(new VectorFEDomainLFIntegrator(f_vec_coeff)); }
+   if (L2)
+   {
+      // DG boundary conditions are enforced weakly with this integrator.
+      b.AddBdrFaceIntegrator(new DGDirichletLFIntegrator(u_coeff, -1.0, kappa));
+   }
+   b.Assemble();
+
+   GridFunction x(&fes);
+   if (H1 || L2) { x.ProjectCoefficient(u_coeff);}
+   else { x.ProjectCoefficient(u_vec_coeff); }
+
+   Vector X, B;
+   OperatorHandle A;
+   a.FormLinearSystem(ess_dofs, x, b, A, X, B);
+
+#ifdef MFEM_USE_SUITESPARSE
+   LORSolver<UMFPackSolver> solv_lor(a, ess_dofs);
+#else
+   LORSolver<GSSmoother> solv_lor(a, ess_dofs);
+#endif
+
+   CGSolver cg;
+   cg.SetAbsTol(0.0);
+   cg.SetRelTol(1e-12);
+   cg.SetMaxIter(500);
+   cg.SetPrintLevel(1);
+   cg.SetOperator(*A);
+   cg.SetPreconditioner(solv_lor);
+   cg.Mult(B, X);
+
+   a.RecoverFEMSolution(X, b, x);
+
+   double er =
+      (H1 || L2) ? x.ComputeL2Error(u_coeff) : x.ComputeL2Error(u_vec_coeff);
+   cout << "L2 error: " << er << endl;
+
+   // Save the solution and mesh to disk. The output can be viewed using GLVis
+   // as follows: "glvis -m mesh.mesh -g sol.gf"
+   x.Save("sol.gf");
```

Comment:
Perhaps add `if (visualization)`, so the user can decide whether to produce these files. Since ParaView is more rarely used than glvis (at least by me), it may also be convenient to have a ParaView output parameter, false by default, but I don't feel strongly about it.